Dependency Extraction Webpack Plugin: Add Module support

sirreal opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?

Add support for extracting modules information with the Dependency Extraction Webpack Plugin.

Part of #57492 

## Why?

As the [modules support matures](https://github.com/WordPress/gutenberg/issues/55942), we should provide at least the same experience folks currently have with scripts.

## How?

First: Behavior with scripts remains unchanged. Folks can continue to use this plugin in exactly the same way without any changes. These changes should be low risk and the module behavior can be considered experimental. **Should we add a note about that in the README?**

We add the necessary functionality to this plugin to support modules.

To determine whether to compile in module or script mode, we check the webpack option `output.modules`. If this option is true, the plugin will compile for modules.

A new option is added `requestToExternalModule`, it serves the same purpose as `requestToExternal` for scripts. The same option could have been reused, but I created a new option to avoid confusion.

One notable difference is that, we extract information about whether a module is dynamic `const x = await import( … )` or static `import x from …` so `wp_register_module` can correctly list static and dynamic module dependencies.

## Testing Instructions

The test suite has been adapted to produce module output. You can check the snapshots added in the PR.

sirreal committed:
Changelog placeholder

sirreal committed:
Implement Module API output

sirreal commented on packages/dependency-extraction-webpack-plugin/lib/.eslintrc.json:
```diff
@@ -1,6 +1,4 @@
 {
-	// Use the default eslint parser. Prevent babel transforms.
-	"parser": "espree",
```
This prevented us from using optional chaining.

sirreal commented on packages/dependency-extraction-webpack-plugin/lib/index.js:
```diff
@@ -32,27 +35,56 @@ class DependencyExtractionWebpackPlugin {
 		);
 	}
 
+	// eslint-disable-next-line jsdoc/check-line-alignment
+	/**
+		* @type {((
+			data: ExternalItemFunctionData,
+			callback: (
+				err?: null | Error,
+				result?: string | boolean | string[] | { [index: string]: any }
+			) => void
+	  ) => void)} */
```
```suggestion
```

sirreal commented on packages/dependency-extraction-webpack-plugin/lib/index.js:
```diff
@@ -32,27 +35,56 @@ class DependencyExtractionWebpackPlugin {
 	externalizeWpDeps( { request }, callback ) {
 		let externalRequest;
 
+		if ( this.useModules ) {
+			// Handle via options.requestToExternal first.
```
```suggestion
		// Handle via options.requestToExternal(Module)  first.
		if ( this.useModules ) {
```

sirreal commented on packages/dependency-extraction-webpack-plugin/lib/index.js:
```diff
@@ -32,27 +35,56 @@ class DependencyExtractionWebpackPlugin {
+					this.options.requestToExternalModule( request );
+			}
+		}
 		// Handle via options.requestToExternal first.
```
```suggestion
```

sirreal committed:
Update snapshots

sirreal committed:
Bundle interactivity

This package is not suitable to be used as a script dependency

sirreal committed:
Update snaps

sirreal committed:
Update README about module handling

sirreal committed:
Add type => module to module asset files

sirreal commented on packages/dependency-extraction-webpack-plugin/lib/index.js:
```diff
@@ -104,7 +138,14 @@ class DependencyExtractionWebpackPlugin {
 	apply( compiler ) {
+		this.useModules = Boolean( compiler.options.output?.module );
+
+		if ( this.useModules ) {
+			this.externalsPlugin.type = 'module';
+		}
```
This isn't necessary at all, but it may be helpful for folks to know whether they've compiled for modules or not.

Otherwise, the output could be identical. Webpack can be very difficult to debug, so seeing this in the output may be helpful.

sirreal commented on packages/dependency-extraction-webpack-plugin/lib/util.js:
```diff
@@ -1,10 +1,11 @@
 const BUNDLED_PACKAGES = [
+	'@wordpress/dataviews',
 	'@wordpress/icons',
+	'@wordpress/interactivity',
```
I sorted this list and added `@wordpress/interactivity`. Because it's compiled as a module, it isn't suitable for use as an external script and should be bundled in scripts, not externalized.

gziolo replied:
`@wordpress/interactivity` gets bundled in the view files with these changes if I [follow it correctly](https://github.com/WordPress/gutenberg/pull/57199#issuecomment-1866819787):
<img width="814" alt="Screenshot 2024-01-02 at 09 46 47" src="https://github.com/WordPress/gutenberg/assets/699132/b4bfedc6-cf51-4212-bab2-1f81a1a36a3c">

sirreal replied:
I'll remove the change to add this. It's unrelated.

sirreal replied:
Done.

sirreal commented on packages/dependency-extraction-webpack-plugin/test/fixtures/wordpress-interactivity/index.js:
```diff
@@ -0,0 +1,13 @@
+import _ from 'lodash';
+
+// This module should be externalized
+// import { store } from '@wordpress/interactivity';
```
```suggestion
```

sirreal committed:
Remove commented line

sirreal committed:
Fix dynamic test

sirreal committed:
Check for matching name

Matching the same class is fragile

sirreal commented on packages/dependency-extraction-webpack-plugin/CHANGELOG.md:
```diff
@@ -4,8 +4,12 @@
+
+### New Features
+
+-   The plugin now supports generating ECMAScript modules output **@TODO @sirreal complete this**.
```
```suggestion
-   Add support for producing module-compatible asset files ((#57199)[https://github.com/WordPress/gutenberg/pull/57199])
```

sirreal committed:
Update CHANGELOG entry

sirreal committed:
Fix markdown link syntax

sirreal committed:
Missing period

sirreal commented on packages/dependency-extraction-webpack-plugin/lib/index.js:
```diff
@@ -209,11 +286,20 @@ class DependencyExtractionWebpackPlugin {
 				version: contentHash,
 			};
 
+			if ( this.useModules ) {
+				assetData.type = 'module';
+			}
```
It's not necessary to add this at the moment, but it may be helpful to know this is a module asset file.

gziolo replied:
Would it make sense to reuse `register_block_script_handle` for the `module` type and branch code depending on this setting in the asset data?

sirreal replied:
That's exactly the kind of thing where `{"type": "module"}` would be helpful 👍 

I'm not aware of any Modules work that mentions changes to the assets file (aside from the changes to the dependencies array implied by https://github.com/WordPress/wordpress-develop/pull/5818 and https://github.com/WordPress/gutenberg/pull/57231). The only work I'm aware of there is #57437.

sirreal replied:
Do you know if there's a specification of asset files somewhere? I want to make sure we update that accordingly.

sirreal committed:
Revert addition of @wordpress/interactivity to bundled scripts

sirreal committed:
Update snapshots

luisherranz commented on packages/dependency-extraction-webpack-plugin/README.md:
```diff
@@ -88,6 +97,66 @@ By default, the following module requests are handled:
+### Behavior with modules
+
+This section describes the behavior of this package to bundle ECMAScript modules and generate asset
+files suitable for use with the WordPress Module API.
```
Should it be `WordPress Module API` or `WordPress Modules API`? I've been using the plural (`Modules`) because the script's one is usually referred to as `WP Scripts`. But we can still change it if the singular makes more sense.

luisherranz commented on packages/dependency-extraction-webpack-plugin/README.md:
```diff
@@ -56,7 +60,12 @@ const config = {
+### Behavior
+
+**Note**: This section describes the classic behavior with webpack and WordPress scripts.
+For information about usage with modules, jump to the [behavior with modules](#behavior-with-modules) section.
```
Maybe instead of the note, we could just say `### Behavior with scripts` and `### Behavior with modules`.

luisherranz commented on packages/dependency-extraction-webpack-plugin/README.md:
````diff
@@ -233,6 +341,19 @@ $script_url = plugins_url( $script_path, __FILE__ );
+$module_url = plugins_url( $module_path, __FILE__ );
+wp_register_module( 'my-module', $module_url, $module_asset['dependencies'], $module_asset['version'] );
+wp_enqueue_module( 'my-module' );
+```
````
Minor corrections.

````suggestion
```php
$module_path       = 'path/to/module.js';
$module_asset_path = 'path/to/module.asset.php';
$module_asset      = file_exists( $module_asset_path )
	? require( $module_asset_path )
	: array( 'dependencies' => array(), 'version' => filemtime( $module_path ) );
$module_url = plugins_url( $module_path, __FILE__ );
wp_register_module( 'my-module', $module_url, $module_asset['dependencies'], $module_asset['version'] );
wp_enqueue_module( 'my-module' );
```
````

luisherranz commented on packages/dependency-extraction-webpack-plugin/lib/util.js:
```diff
@@ -56,6 +56,21 @@ function defaultRequestToExternal( request ) {
+ *   to ignore the request. Return `string` to map the request to an external. This may simply be returning the request, e.g. `@wordpress/interactivity` maps to the external `@wordpress/interactivity`.
+ */
+function defaultRequestToExternalModule( request ) {
+	if ( request === '@wordpress/interactivity' ) {
+		return request;
+	}
```
This is where maybe we should throw if people try to add other `@wordpress` packages, right?

sirreal replied:
Maybe, but I'd prefer to handle that in another PR.

luisherranz replied:
Sure 🙂

luisherranz commented on packages/dependency-extraction-webpack-plugin/test/__snapshots__/build.js.snap:
```diff
@@ -1,5 +1,230 @@
+`;
+
+exports[`DependencyExtractionWebpackPlugin modules Webpack \`dynamic-import\` should produce expected output: Asset file 'main.asset.php' should match snapshot 1`] = `
+"<?php return array('dependencies' => array('@wordpress/blob'), 'version' => '092c2bce8c247ee11100', 'type' => 'module');
```
A static dependency of a dynamic dependency should also be considered dynamic. 

```js
// index.js
import( './util' ).then( () => {} );
```

```js
// util.js
import { isBlobURL } from '@wordpress/blob';
```

In this case, if we mark `@wordpress/blob` as a static dependency, we will preload it when `index.js` is enqueued, which is not correct. It's true that when `index.js` finally imports `util.js`, it results in a waterfall, but that's the way modules work 🤷‍♂️

luisherranz reviewed (changes requested):
This is going great! 👏🙂

I think the only change we need is to mark the static dependencies of the dynamic modules as dynamic instead of static. The `dynamic-import` test is a clear example of a static dependency that should not be preloaded because it's behind a dynamic `import()`.

It's tricky because that dynamic `import("./util.js")` of the `index.js` file is not a native module `import()`, but handled internally by Webpack. For native dynamic `import()`s, we already manage that in the Modules API (I just [added a test](https://github.com/WordPress/wordpress-develop/pull/5818/commits/ef978d8fb876f45e5449dcd3a0ea6c8540700deb)  for that use case).

It's quite an edge case, but do you think it's doable?

sirreal committed:
Apply README suggestions, add experimental warning

sirreal committed:
Fix README modules PHP example

Co-authored-by: Luis Herranz <luisherranz@gmail.com>

sirreal committed:
Search for a static path to root to test for dynamic

sirreal committed:
Add name to webpack config for easier test debugging

sirreal committed:
Update snapshots

sirreal committed:
Add cyclic static dep test

sirreal committed:
Add cyclic dynamic test

sirreal commented:
Thanks for flagging that issue with the dynamic checking @luisherranz! I pushed some changes to fix the static/dynamic check for dependencies. For a dependency to be static, we should be able to trace static imports from the entrypoint to the module. If that sequence is ever broken by a dynamic import, the dependency is dynamic.

This involved module graph traversal. I added some tests for circular dependencies because I was a bit worried we could get into an infinite loop, but there's no issue with the tests I added.

jsnajdr commented on packages/dependency-extraction-webpack-plugin/lib/util.js:
```diff
@@ -56,6 +56,21 @@ function defaultRequestToExternal( request ) {
+ */
+function defaultRequestToExternalModule( request ) {
+	if ( request === '@wordpress/interactivity' ) {
+		return request;
```
If here I return something else than `request`, for example, a different (mapped) module name, will the plugin transform the `import '@wordpress/interactivity'` to `import '@wordpress/mapped-interactivity'`?

And will it output the _mapped_ name to the `.asset.php` file? I suspect it will output the original one.

The "scripts" version has the `requestToHandle` function to produce the handles that go into the `.asset.php` file.

sirreal replied:
You're right, that's a bug. If we map to a different name, the original is still included in the asset file. I'll address it and add a test.

I don't think we need both functions for modules (toExternal and toHandle), we should be able to use a single module ID for the external module import and the WordPress module ID.

I'd even consider changing the signature (and name of the option) so that it expects to return `true` to indicate we want to externalize the module and track the dependency. It doesn't seem like we need to be rewriting the imports.

Any thoughts on that?

sirreal replied:
After some more reflection, there are likely use cases for rewriting module names.

For example, maybe I want to include a specific version of a package on NPM, but I'd like to rename it in the WordPress module system to avoid collisions. I can use the regular import with the npm name in my source files, but rewrite to a unique name for the external and the WordPress module ID.

I still don't see the need for two functions (toExternal and toHandle), one seems sufficient.

jsnajdr replied:
> I still don't see the need for two functions (toExternal and toHandle), one seems sufficient.

That's fine, I agree that for modules we need only one. There are only module names (like `@wordpress/interactivity`), nothing else. With `scripts` there are two distinct entities: the `window.wp.blob` global variable produced by `toExternal` and the `wp-blob` WP script handle produced by `toHandle`.

Taking care to add the mapped value produced by `toExternal` to `.asset.php`, not the original one, is enough to address my concerns.

> After some more reflection, there are likely use cases for rewriting module names.

There could be also a `true` return value that says "externalize without changing the name". Equivalent to returning the `request` value.

sirreal replied:
I pushed a test and a fix for the renamed module bug. Thanks! I don't plan to implement the `true` shorthand to use the same module ID for now, but it's easy enough if we'd like to add it at some point.

jsnajdr replied:
> I don't plan to implement the `true` shorthand to use the same module ID for now

I think it's a good idea to have this shorthand. The evidence is that literally all use cases we currently have, both in production and it tests, could use the `true` return value. Another thing to add to my todo 🙂

sirreal replied:
I can make a quick PR for this.

sirreal replied:
https://github.com/WordPress/gutenberg/pull/57593

jsnajdr commented on packages/dependency-extraction-webpack-plugin/README.md:
```diff
@@ -88,6 +94,68 @@ By default, the following module requests are handled:
+
+**Note:** This plugin overlaps with the functionality provided by [webpack `externals`](https://webpack.js.org/configuration/externals). This plugin is intended to extract module handles from bundle compilation so that a list of module dependencies does not need to be manually maintained. If you don't need to extract a list of module dependencies, use the `externals` option directly.
+
+This plugin is compatible with `externals`, but they may conflict. For example, adding `{ externals: { '@wordpress/blob': 'wp.blob' } }` to webpack configuration will effectively hide the `@wordpress/blob` module from the plugin and it will not be included in dependency lists.
```
It's more precise to say not that the plugin "overlaps" with `externals`, but that it actually uses the `externals` plugin internally! It creates a configuration for it and passes it to the `externals` plugin to do the actual work.

A conflict may arise when you create another instance of the externals plugin in your config, and they both want to process the same module. Then only one can win.

The other, fairly independent functionality of DEWP is finding all the external imports in the bundle and recording them in the `.asset.php` file.

The way how both functionalities are mixed together can be confusing for users who are trying to understand what DEWP does. We discussed that once in 2022 with @anomiex but never took any real action to fix it.

sirreal replied:
This part is copied from the "behavior with scripts" section. There's certainly nuance here, but `externals` in this case refers to [`externals` as a configuration option](https://webpack.js.org/configuration/externals/#externals) (which I believe uses an `ExternalPlugin` under the hood, but that's not apparent via webpack config).. The goal is to make it clear that you really shouldn't be adding your own externals declarations for WordPress script/module dependencies.

If you have concrete suggestions for how to improve this (and the text in the other section) I'm happy to make changes.

jsnajdr replied:
> If you have concrete suggestions for how to improve this

I think I'll do my own PR with a docs update, especially now when this one is merged :+1:

jsnajdr commented on packages/dependency-extraction-webpack-plugin/lib/index.js:
```diff
@@ -32,27 +35,48 @@ class DependencyExtractionWebpackPlugin {
+		 * Offload externalization work to the ExternalsPlugin.
+		 */
 		this.externalsPlugin = new webpack.ExternalsPlugin(
 			'window',
```
This `'window'` parameter says that the externals are on the `window` global object. Makes sense only for the `useModules = false` mode. Does the parameter do something in `useModules = true` mode or is it completely ignored?

sirreal replied:
This plugin reads the compiler options to know whether to use modules. We don't have access to the compiler when we create an instance of the plugin, only when webpack calls `apply`. At that time we update `useModules` and also set the externals plugin `type` appropriately before calling `apply` on it:

https://github.com/WordPress/gutenberg/blob/0b094345fcc7875c8504dcd4d0829e75ef6981c7/packages/dependency-extraction-webpack-plugin/lib/index.js#L142-L149

jsnajdr replied:
Oh, that's insidious! Is there any way to postpone the creation of `this.externalsPlugin` until we know all the parameters?

sirreal replied:
Sure, we can instantiate the externals plugin inside the `apply` method.

sirreal replied:
I pushed this change.

sirreal committed:
Add test for renamed modules, fix renamed module bug

sirreal committed:
Move ExternalsPlugin instantiation to apply method

Avoid instantiating with one externals type and switching later

luisherranz reviewed (approved):
Thanks, Jon. Everything looks great to me now 🙂

gziolo commented on packages/dependency-extraction-webpack-plugin/README.md:
```diff
@@ -119,7 +187,7 @@ The filename for the generated asset file. Accepts the same values as the Webpac
 -   Default: `false`
 
-By default, one asset file is created for each entry point. When this flag is set to `true`, all information about assets is combined into a single `assets.(json|php)` file generated in the output directory.
+By default, one asset file is created for each entry point. When this flag is set to `true`, all information about assets is combined into a single `asset.(json|php)` file generated in the output directory.
```
That wasn’t a typo. See 

https://github.com/WordPress/gutenberg/blob/d362529ecc78ae3054e1ee65895b8725020dd0c6/packages/dependency-extraction-webpack-plugin/lib/index.js#L250

sirreal replied:
~I think that code has a typo, this outputs `asset.*` by default:~

~All the core blocks use `*.asset.php` and its what core looks for:~

https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-includes/blocks/file/view.asset.php

https://github.com/WordPress/wordpress-develop/blob/e43275b61c6e3a8d5657b6d21eb99bf1bafd52b2/src/wp-includes/blocks.php#L159

sirreal replied:
Oh, I see 🤦 This is specifically for a single asset file containing multiple assets.

I'll revert this change.

sirreal committed:
This is supposed to say "assets"
